Modules/pyexpat.c

```diff
@@ -1650,16 +1650,93 @@ add_submodule(PyObject *mod, const char *fullname)
+    {"XML_ERROR_UNCLOSED_CDATA_SECTION", "unclosed CDATA section"},
+    {"XML_ERROR_EXTERNAL_ENTITY_HANDLING", "error in processing external entity reference"},
+    {"XML_ERROR_NOT_STANDALONE", "document is not standalone"},
+    {"XML_ERROR_UNEXPECTED_STATE", "unexpected parser state - please send a bug report"},
+    {"XML_ERROR_ENTITY_DECLARED_IN_PE", "entity declared in parameter entity"},
+    {"XML_ERROR_FEATURE_REQUIRES_XML_DTD", "requested feature requires XML_DTD support in Expat"},
+    {"XML_ERROR_CANT_CHANGE_FEATURE_ONCE_PARSING", "cannot change setting once parsing has begun"},
+
+    /* Added in Expat 1.95.7. */
+    {"XML_ERROR_UNBOUND_PREFIX", "unbound prefix"},
+
+    /* Added in Expat 1.95.8. */
+    {"XML_ERROR_UNDECLARING_PREFIX", "must not undeclare prefix"},
+    {"XML_ERROR_INCOMPLETE_PE", "incomplete markup in parameter entity"},
+    {"XML_ERROR_XML_DECL", "XML declaration not well-formed"},
+    {"XML_ERROR_TEXT_DECL", "text declaration not well-formed"},
+    {"XML_ERROR_PUBLICID", "illegal character(s) in public id"},
+    {"XML_ERROR_SUSPENDED", "parser suspended"},
+    {"XML_ERROR_NOT_SUSPENDED", "parser not suspended"},
+    {"XML_ERROR_ABORTED", "parsing aborted"},
+    {"XML_ERROR_FINISHED", "parsing finished"},
+    {"XML_ERROR_SUSPEND_PE", "cannot suspend in external parameter entity"},
+
+    /* Added in 2.0.0. */
+    {"XML_ERROR_RESERVED_PREFIX_XML", "reserved prefix (xml) must not be undeclared or bound to another namespace name"},
+    {"XML_ERROR_RESERVED_PREFIX_XMLNS", "reserved prefix (xmlns) must not be declared or undeclared"},
+    {"XML_ERROR_RESERVED_NAMESPACE_URI", "prefix must not be bound to one of the reserved namespace names"},
+
+    /* Added in 2.2.1. */
+    {"XML_ERROR_INVALID_ARGUMENT", "invalid argument"},
+
+    /* Added in 2.3.0. */
+    {"XML_ERROR_NO_BUFFER", "a successful prior call to function XML_GetBuffer is required"},
+
+    /* Added in 2.4.0. */
+    {"XML_ERROR_AMPLIFICATION_LIMIT_BREACH", "limit on input amplification factor (from DTD and entities) breached"}
+};
+
 static int
 add_error(PyObject *errors_module, PyObject *codes_dict,
-          PyObject *rev_codes_dict, const char *name, int value)
+          PyObject *rev_codes_dict, size_t error_index)
 {
-    const char *error_string = XML_ErrorString(value);
+    const char * const name = error_info_of[error_index].name;
+    const char * const error_string = error_info_of[error_index].description;
+    const long error_code = (long)error_index;
+
+#ifndef NDEBUG
+    /* For all codes that the (potentially outdated) runtime libexpat knows
+     * already, its return from function XML_ErrorString should agree with
+     * our copy of error message */
+    const char * const runtime_error_string = XML_ErrorString((int)error_code);
+    assert((runtime_error_string == NULL) || ! strcmp(runtime_error_string, error_string));
+#endif
```

Comment:
So, if you feel like asserting this (in debug mode), why not just use the runtime provided string, if available?

The error strings are owned by libexpat, not by CPython. IMHO, it's ok to provide a fallback so that we can always provide such a string, even if the runtime libexpat does not know it – although that seems somewhat useless since that error will then also not occur. But we should otherwise rely on libexpat to (potentially) know things better than we do, and ask it.

Reply:
Hi @scoder,

what would be the advantage of making `XML_ErrorString` the runtime source of truth?
Since pyexpat passes a _static_ set of integers, is there any gain in information?

If there is any concern about the strings changing in libexpat in the future (of which I have no plans), we should not go store a copy in pyexpat. If we do store a copy, pyexpat becomes the new source of truth.  Everyone running libexpat 2.2.x in the RHEL 8 with six of these error strings only available in pyexpat would probably agree that pyexpat has become the new source of truth.

Maybe the debugging code block was a bit out of place anyway. I have dropped it now, and would like to present the result as my current recommendation.

Best, Sebastian

Reply:
PS: @scoder I realize now that our difference in view above is rather minor. I have implemented your preference now, or so I think. If you change your mind, I have backup branches to restore to, easily :beers:

Reply:
> If there is any concern about the strings changing in libexpat in the future (of which I have no plans), we should not go store a copy in pyexpat.

Well, `libexpat` _is_ the source of truth. Whatever we do in `pyexpat` should follow (unless there is a compelling reason not to).

That being said, I admit that the original design is a bit surprising here. What you would normally expect is an error code as a key to an error message, whether that message is part of the originally reported error "thing" (read, exception) itself or not. Here, the error _message_ is the key to the error code. Obviously, that is not something we can change at this point.

Now, even if there are no plans to change error messages, I'm sure there's precedence for such changes in other projects, so I wouldn't rule it out entirely. And it's not impossible that some users persisted errors in one way or another, so changing messages might break someone's data. But – I think that is not a decision for `pyexpat` to take but by `libexpat`, so:

> If we do store a copy, pyexpat becomes the new source of truth. Everyone running libexpat 2.2.x in the RHEL 8 with six of these error strings only available in pyexpat would probably agree that pyexpat has become the new source of truth.

I don't think so. Old versions of libexpat do not change any more, and we know their behaviour. We can happily provide fallback constants for them (especially since they won't trigger them). `pyexpat` should not become a source of truth at all, but it can become a source of backported functionality. That allows `pyexpat` users to write simple code that responds to specific errors even if they won't occur in the current runtime `libexpat` version.

Reply:
> PS: @scoder I realize now that our difference in view above is rather minor. I have implemented your preference now, or so I think. If you change your mind, I have backup branches to restore to, easily beers

Thanks. Yes, I think that's how it should work.

Reply:
@scoder I think we have different meanings for "source of truth" in mind. Beyond that, I don't we disagree. I'm optimistic that my latest changes already do what you wanted. Let me know.